README.md
#FastExcel
FastExcel is a FAST and TINY excel read/write component.It's FAST and TINY.

FastExcel is content-based,that means we just care about the content of excel. So FastExcel just read the cell string and other important information,Some Properities like color,font are not supported.Because we don't need to read,parse,store these additional information so FastExcel just need little memory.

- Reading and Writing support for Excel '97(-2003) (BIFF8) file format.
- Low level structures for BIFF(Binary Interchange File Format).
- Low level structures for compound document file format (also known as "OLE2 storage file format" or "Microsoft - Office compatible storage file format").
- API for creating, reading excel file.

Lastest Version 0.5.1	2010-12-25
##DOWNLOAD
FastExcel 0.5.1 Release

[DOWNLOAD][release]
##Example
**Basic Read**
```java
	public void testDump() throws ExcelException {
		Workbook workBook;
		workBook = FastExcel.createReadableWorkbook(new File("test.xls"));
		workBook.setSSTType(BIFFSetting.SST_TYPE_DEFAULT);//memory storage
		workBook.open();
		Sheet s;
		s = workBook.getSheet(0);
		System.out.println("SHEET:"+s);
		for (int i = s.getFirstRow(); i < s.getLastRow(); i++) {
		System.out.print(i+"#");
		for (int j = s.getFirstColumn(); j < s.getLastColumn(); j++) {
			System.out.print(","+s.getCell(i, j));
		}
		System.out.println();
		}
		workBook.close();
	}
```
**Event-based Read**
```java
  public void testEventDump() throws ExcelException {
		Workbook workBook;
		workBook = FastExcel
				.createReadableWorkbook(new File("test.xls"));
		workBook.open();
		workBook.getSheet(0, new SheetReadAdapter() {
			public void onCell(int row, int col, String content) {
				System.out.println(row + "," + col + "," + content);
			}
		});
		workBook.close();
	}
```
**Basic Write**
```java
  public void testWrite() throws ExcelException{
		File f=new File("write.xls");
		Workbook wb=FastExcel.createWriteableWorkbook(f);
		wb.open();
		Sheet sheet=wb.addSheet("sheetA");
		sheet.setCell(1, 2, "some string");
		wb.close();
	}
	
```
**Stream Write**
```java
  public void testStreamWrite() throws Exception{
		File f=new File("write1.xls");
		Workbook wb=FastExcel.createWriteableWorkbook(f);
		wb.open();
		Sheet sheet=wb.addStreamSheet("SheetA");
		sheet.addRow(new String[]{"aaa","bbb"});
		wb.close();
	}
```


##How it works
A workbook document with several sheets (BIFF5-BIFF8) is usually stored using the compound document file format (also known as "OLE2 storage file format" or "Microsoft Office compatible storage file format"). It contains several streams for different types of data.Depending on the document type, different names are used for the stream(s) they contain. BIFF5-BIFF8 workbook documents that are stored in a compound document file contain a stream in the root storage called the Workbook Stream. The name of this stream in the compound document file is "Book" for BIFF5 workbooks, and "Workbook" for BIFF8 workbooks.If a BIFF5-BIFF8 workbook document is stored as stream file, the entire stream is called the Workbook Stream. In BIFF5-BIFF8 Workbook Streams, the Workbook Globals Substream is not the leading part of the stream. It is followed by all Sheet Substreams in order of the sheets that are in the document. Common structure of a BIFF5-BIFF8 Workbook Stream:

- Workbook Globals Substream (required)
- First Sheet Substream (required)
- Second Sheet Substream (optional)
- Third Sheet Substream (optional)

Most of the Excel streams or substreams are divided into records. Each record contains specific data for the various contents or features in a document. It consists of a header specifying the record type and size, followed by the record data. Common structure of a BIFF record:


Offset  |Size	  |Contents
:-------|:-----:|:-------
0	    |2   	|Identifier
2	    |2  	|Size of the following data (sz)
4	    |sz  	|Record data

To read excel file.FastExcel parses these records and build an inner struct of excel file.The Record Parsers:

- BOFParser
- EOFParser
- BoundSheetParser
- SSTParser
- IndexParser
- DimensionParser
- RowParser
- LabelSSTParser
- RKParser
- MulRKParser
- LabelParser
- BoolerrParser
- XFParser
- FormatParser
- DateModeParser
- StyleParser
- MulBlankParser
- NumberParser
- RStringParser

##What's New
add formula support.

##License
It runs on the Java 2 Platform (JDK 1.4 or later) .FastExcel is licensed under the terms of the GNU Lesser General Public Licence (LGPL).  A copy of the licence is included in the distribution.

###Contract 

- yAma <guooscar@gmail.com>
- HeDYn <hedyn@foxmail.com>
- 千橙 <im.qiancheng@gmail.com>

###References
1. <http://en.wikipedia.org/wiki/Microsoft_Excel>
2. <http://sc.openoffice.org/excelfileformat.pdf>
3. <http://sc.openoffice.org/compdocfileformat.pdf>
4. <http://blog.csdn.net/liangjingbo/archive/2008/09/03/2874959.aspx>

[release]: https://github.com/iqiancheng/FastExcel/releases


changelog.md
FastExcel change log
NOTE:
	the '*' means change
	the '+' means add
	the '-' means remove

--------------------------------------------------------------------------------
CHANGE IN VERSION 0.5.1  2010-12-25 HeDYn<hedyn@foxmail.com>		
--------------------------------------------------------------------------------	
edu.npu.fastexcel.compound.DirectoryEntry 
中增加 String WORKBOOK_NAME_OTHER = "Book"; 
另解决了读取到的数值型结果的公式值与显示的不一致的问题。
与0.5相比修改了以下几个类
edu.npu.fastexcel.compound.DirectoryEntry
edu.npu.fastexcel.biff.read.BIFFReader
edu.npu.fastexcel.biff.read.WorkBookStream
edu.npu.fastexcel.biff.parser.cell.FormulaParser
edu.npu.fastexcel.biff.parser.globals.Format
 
--------------------------------------------------------------------------------
CHANGE IN VERSION 0.5  2010-12-2 HeDYn<hedyn@foxmail.com>		
--------------------------------------------------------------------------------	
1.add formula support HeDYn<hedyn@foxmail.com>		

--------------------------------------------------------------------------------
CHANGE IN VERSION 0.4.4  2010-1-10 yAma<guooscar@gmail.com>
--------------------------------------------------------------------------------
1.fix bug: that when writing excel,the dimension 's last row is wrong.
thanks to 高雨峰 <sillyfox008@sina.com>
	
	
--------------------------------------------------------------------------------
CHANGE IN VERSION 0.4.3  2009-11-5 yAma<guooscar@gmail.com>
--------------------------------------------------------------------------------
1.fix bug:date format and number format error.

--------------------------------------------------------------------------------
CHANGE IN VERSION 0.4.2  2009-5-19 yAma<guooscar@gmail.com>
--------------------------------------------------------------------------------
1.fix bug:ClassCastException when open workbookGlobalStream
	Thanks to maojianke001@pingan.com.cn

--------------------------------------------------------------------------------
CHANGE IN VERSION 0.4.1  2009-5-13 yAma<guooscar@gmail.com>
--------------------------------------------------------------------------------	
1.fix bug:can not delete file after workbook closed.
  Thanks to allan <allan@dreamtech.com.cn>
  
  
--------------------------------------------------------------------------------
CHANGE IN VERSION 0.4  2009-4-26 yAma<guooscar@gmail.com>
--------------------------------------------------------------------------------	
1.add SST,ArraySST,FileSST,CacheFileSST
2.change DefaultSheetStream
3.add setSSTType in BIFFSetting
4.add getFirstColIndex and getLastColIndex method

--------------------------------------------------------------------------------
CHANGE IN VERSION 0.3.4  2009-4-22 yAma<guooscar@gmail.com>
--------------------------------------------------------------------------------	
1.fix bug:setup wrong dimension when writing excel.
2.fix bug:open generated excel using Microsoft Excel,change nothing,save it,open
again. 
-edu.npu.fastexcel.biff.write.SheetStreamWriter
*edu.npu.fastexcel.biff.write.FileSheetStreamWriter
*edu.npu.fastexcel.biff.write.DefaultSheetStreamWriter
*edu.npu.fastexcel.biff.write.WorkBookGlobalsStreamWriter

--------------------------------------------------------------------------------
CHANGE IN VERSION 0.3.4  2009-3-31  yAma<guooscar@gmail.com>
--------------------------------------------------------------------------------	
1.add a stream write interface.	
	
--------------------------------------------------------------------------------
CHANGE IN VERSION 0.3.3  2009-3-18  yAma<guooscar@gmail.com>
--------------------------------------------------------------------------------
1.fix bug:NumUtil.getRKValue.Thanks to maojianke001@pingan.com.cn		

--------------------------------------------------------------------------------
CHANGE IN VERSION 0.3.2  2009-2-25  yAma<guooscar@gmail.com>
--------------------------------------------------------------------------------
1.add date format testing	
	
--------------------------------------------------------------------------------
CHANGE IN VERSION 0.3.1  2009-2-23  yAma<guooscar@gmail.com>
--------------------------------------------------------------------------------
1.add format 22(yyyy-M-d HH:mm)
2.add user defined format interface
	
	
--------------------------------------------------------------------------------
CHANGE IN VERSION 0.3  2009-2-17  yAma<guooscar@gmail.com>
--------------------------------------------------------------------------------
1.add write interface
2.fix some bugs of compound reader
3.fix some bugs of date format and number format

--------------------------------------------------------------------------------
CHANGE IN VERSION 0.2  2009-1-16  yAma<guooscar@gmail.com>
--------------------------------------------------------------------------------
1.add a event-based read interface.
	*edu.npu.fastexcel.ReadableWorkbook
	+edu.npu.fastexcel.SheetReadAdapter
	+edu.npu.fastexcel.SheetReadListener
2.add ant build file
3.add change log file
4.add license declare

	

src/edu/npu/fastexcel/FastExcel.java
/* 
 *FastExcel,(c) copyright 2009 yAma<guooscar@gmail.com>.  
 *WEB: http://fastexcel.sourceforge.net
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2.1 of the License, or
 * (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
 * USA.
 */
package edu.npu.fastexcel;

import java.io.File;

/**
 * <a href="http://fastexcel.sourceforge.net">FastExcel</a>:<strong>FAST</strong>
 * and <strong>TINY</strong> excel read/write component.<br/> 
 * change log:
 * <ul>
 * <li>create file 2008-12-26</li>
 * <li>version 0.1 first release 2009-1-1</li>
 * <li>version 0.2 add event-based interface. 2009-1-16</li>
 * <li>version 0.3 add write interface.2009-2-16</li>
 * <li>version 0.3.1 add user define format interface.2009-2-23</li>
 * <li>version 0.3.2 add date format testing.2009-2-25</li>
 * <li>version 0.3.3 fix bug:NumUtil.getRKValue.
 * Thanks to maojianke001@pingan.com.cn. 2009-3-18</li>
 * <li>version 0.3.4 add a stream write interface.2009-3-31</li>
 * <li>version 0.3.5 fix bugs.Thanks XUJIANFENG004@pingan.com.cn.2009-4-22</li>
 * <li>version 0.4
 * 		<ol>add SST,ArraySST,FileSST,CacheFileSST</ol>
 *		<ol>change DefaultSheetStream</ol>
 *		<ol>add setSSTType in BIFFSetting</ol>
 *	2009-4-26
 * </li>
 * <li>version 0.4.1 fix bug:can not delete file after workbook closed.
 * Thanks to allan <allan@dreamtech.com.cn>. 2009-5-13</li>
 * <li>version 0.4.2 fix bug:ClassCastException when open workbookGlobalStream
 * Thanks to maojianke001@pingan.com.cn 2009-5-19</li>
 * <li>
 * version 0.4.3 fix bug:date format and number format error. 2009-11-5
 * </li>
 * <li>
 * version 0.4.4
 * fix bug: that when writing excel,the dimension 's last row is wrong.
 * thanks to 高雨峰 <sillyfox008@sina.com> 2010-1-10
 * 
 * version 0.5
 * add formula support HeDYn<hedyn@foxmail.com>
 * </li>
 * </ul>
 * For my love LIMIAO
 * @author <a href="guooscar@gmail.com">yAma</a> 
 * @version 0.4.4
 */
public class FastExcel {
	/**
	 * create a readable workbook from given file
	 * @param file the file name
	 * @return the readable workbook
	 */
	public static Workbook createReadableWorkbook(File file) {
		ReadableWorkbook workbook = new ReadableWorkbook(file);
		return workbook;
	}
	/**
	 * create a writable workbook from given file
	 * @param file the file name
	 * @return the writable workbook
	 */
	public static Workbook createWriteableWorkbook(File file) {
		return new WritableWorkbook(file);
	}
}
